src/Database/Retry/ConnectWaitStrategy.php

```diff
@@ -0,0 +1,87 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         4.1.3
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Database\Retry;
+
+use Cake\Core\Retry\RetryStrategyInterface;
+use Exception;
+use PDOException;
+
+/**
+ * Handles determining if a failed connection should be attempted again.
+ *
+ * @internal
+ */
+class ConnectWaitStrategy implements RetryStrategyInterface
+{
+    /**
+     * @var int[]
+     */
+    protected $errorCodes;
+
+    /**
+     * @var int
+     */
+    protected $maxRetries;
+
+    /**
+     * @var int
+     */
+    protected $retryInterval;
+
+    /**
+     * @param array $errorCodes DB-specific error codes that allow retrying
+     * @param int $maxRetries Maximum number of retry attempts
+     * @param int $retryInterval Number of seconds to wait before retrying
+     */
+    public function __construct(array $errorCodes = [], int $maxRetries = 1, int $retryInterval = 5)
+    {
+        $this->errorCodes = $errorCodes;
+        $this->maxRetries = $maxRetries;
+        $this->retryInterval = $retryInterval;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getMaxRetries(): int
+    {
+        return $this->maxRetries;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function shouldRetry(Exception $exception): bool
+    {
+        if (
+            $exception instanceof PDOException &&
+            (
+                in_array($exception->getCode(), $this->errorCodes) ||
+                (
+                    $exception->errorInfo &&
+                    in_array($exception->errorInfo[1], $this->errorCodes)
```

Comment:
This situation is caused by https://bugs.php.net/bug.php?id=64705. errorInfo is null when the exception is thrown from PDO construction.